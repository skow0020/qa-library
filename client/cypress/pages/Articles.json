{
    "addArticle": "#add-article",
    "firstArticleCard": "#article-card-0",
    "firstArticleCartTitle": "#article-card-0 .card-title",
    "articleCards": "[id*=article-card]",
    "category": "#category",
    "language": "#language",
    "cardPosts": ".card-post"
}