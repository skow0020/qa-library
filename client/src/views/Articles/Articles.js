/* eslint-disable react-hooks/exhaustive-deps */

import React, { useEffect, useState } from 'react';

import Button from '@material-ui/core/Button';
import CardComponent from 'components/common/CardComponent';
import CategoriesSelection from 'components/common/CategoriesSelection';
import Chip from '@material-ui/core/Chip';
import Colors from 'utils/Colors';
import Grid from '@material-ui/core/Grid';
import LanguagesSelection from 'components/common/LanguagesSelection';
import { Link } from 'react-router-dom';
import LoadError from 'components/common/LoadError';
import Loading from 'components/common/Loading';
import PageTitle from 'components/common/PageTitle';
import { getCategoryTheme } from 'utils/util';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(() => ({
  addButton: {
    backgroundColor: Colors.primary,
    color: Colors.white,
    marginLeft: 'auto'
  }
}));

export default function Articles() {
  const classes = useStyles();
  const [articles, setArticles] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [category, setCategory] = useState('');
  const [language, setLanguage] = useState('');
  const [error, setError] = useState('');

  useEffect(() => {
    getArticles();
  }, []);

  useEffect(() => {
    getArticles();
  }, [category]);
  useEffect(() => {
    getArticles();
  }, [language]);

  const getArticles = () => {
    const categoryFilter = category ? `category=${category}` : '';
    const languageFilter = language ? `language=${language}` : '';
    const filter = `${categoryFilter}&${languageFilter}`;

    setIsLoading(true);
    fetch(`/api/articles?${filter}`)
      .then((response) => response.json())
      .then(
        (data) => {
          setArticles(data.data);
          setIsLoading(false);
        },
        (error) => {
          setError(error);
          setIsLoading(false);
        }
      );
  };

  if (isLoading) return <Loading />;

  if (error) return <LoadError error='Articles failed to load' />;

  return (
    <Grid container>
      <Grid container alignItems='center'>
        <PageTitle title='Articles' />
        <Button
          id='add-article'
          component={Link}
          to='/add-article'
          variant='contained'
          className={classes.addButton}
        >
          Add Article
        </Button>
      </Grid>
      <Grid>
        <form id='filtering-form'>
          <CategoriesSelection
            id='category'
            value={category}
            onChange={(e) => setCategory(e.target.value)}
          />
          <LanguagesSelection
            id='language'
            value={language}
            onChange={(e) => setLanguage(e.target.value)}
          />
        </form>
      </Grid>
      <Grid container spacing={4}>
        {articles.map((post, idx) => (
          <Grid item md={4} key={idx}>
            <CardComponent
              idx={`article-card-${idx}`}
              url={post.url}
              urlTarget='_blank'
              title={post.title}
              subheader={`By ${post.author}`}
              avatar={
                <Chip
                  size='small'
                  label={post.category}
                  style={{
                    backgroundColor: post.category
                      ? getCategoryTheme(post.category)
                      : Colors.blue
                  }}
                />
              }
              backgroundImage={post.backgroundImage}
              body={post.body}
            />
          </Grid>
        ))}
      </Grid>
    </Grid>
  );
}
