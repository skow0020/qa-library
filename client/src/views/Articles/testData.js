export const articles = {
  data: [
    {
      '_id': '5e220996d7ca7340bcfc9dd5',
      'backgroundImage': 'https://i-cdn.phonearena.com/images/article/104071-two_lead/Just-for-you-Here-are-10-paid-iOS-apps-free-for-a-limited-time.jpg',
      'author': 'Benny Schaden Jr.',
      'category': 'General',
      'language': 'Swift',
      'url': 'https://qa-library-dev.herokuapp.com/qa-dashboard',
      'title': 'How to sand a hippo',
      'body': 'Est quibusdam laboriosam ipsum cum est sunt numquam voluptas. Quas cupiditate facere vero.',
      'article_id': 1000
    },
    {
      '_id': '5e220996d7ca7340bcfc9dd6',
      'backgroundImage': 'https://i-cdn.phonearena.com/images/article/104071-two_lead/Just-for-you-Here-are-10-paid-iOS-apps-free-for-a-limited-time.jpg',
      'author': 'Joelle Mayer',
      'category': 'UI Automation',
      'language': 'Swift',
      'url': 'https://qa-library-dev.herokuapp.com/qa-dashboard',
      'title': 'How to sand a hippo',
      'body': 'Dolorum consequuntur suscipit impedit impedit repudiandae qui cupiditate quis rerum. Veritatis ab ab culpa praesentium necessitatibus vitae quasi.',
      'article_id': 1001
    },
    {
      '_id': '5e220996d7ca7340bcfc9dd7',
      'backgroundImage': 'https://i-cdn.phonearena.com/images/article/104071-two_lead/Just-for-you-Here-are-10-paid-iOS-apps-free-for-a-limited-time.jpg',
      'author': 'Leonardo Wunsch',
      'category': 'DeVops',
      'language': 'Swift',
      'url': 'https://qa-library-dev.herokuapp.com/qa-dashboard',
      'title': 'How to sand a hippo',
      'body': 'Quas eius non iure. Voluptatibus cupiditate recusandae libero repellendus velit ex.',
      'article_id': 1002
    }
  ]
};