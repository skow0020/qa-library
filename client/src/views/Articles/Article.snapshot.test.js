import Articles from './Articles';
import React from 'react';
import renderer from 'react-test-renderer';

describe('Articles Snapshot Tests', () => {
  test('Articles snapshot', () => {
    const component = renderer.create(
      <Articles></Articles>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});