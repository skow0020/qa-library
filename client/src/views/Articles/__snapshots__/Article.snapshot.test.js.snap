// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles Snapshot Tests Articles snapshot 1`] = `
<div
  className="main-content-container px-4 pb-4 container-fluid"
>
  <div
    className="error"
  >
    <div
      className="error__content"
    >
      <h3>
        Loading...
      </h3>
    </div>
  </div>
</div>
`;
